### PR TITLE
Add prediction of feature contributions

### PR DESCRIPTION
**Pull request by mcuelenaere**
This implements the idea described at http://blog.datadive.net/interpreting-random-forests/
which tries to give insight in how a prediction is composed of its feature contributions
and a bias.

The calculation of the node values is based on the findings of @SauceCat at https://github.com/dmlc/xgboost/issues/1986.

Open issues:
 1. I added this functionality to `XGBoosterPredict()` as another value to `option_mask`. However it is mutually exclusive with the `PredictLeaf()` and `PredictRaw()` options, so it feels wrong to being able to pass `0x07` to this function (which is no different from `0x04`)
 1. to calculate the values at each node, I multiply the learning rate with the `base_weight` property. However, the `learning_rate` value seems to be unavailable in `GBTree` (or `RegTree` for that matter). So the code calculates it by dividing the leaf value with the base weight, which is pretty hacky.
 1. only the Python bindings have been updated to expose this new functionality

### CODE REVIEW

**Comment by khotilov** [comment]
That would be useful to have. I can compare it to my parser that does a similar thing but in a more awkward manner. However, it is usually in contributor's court to demonstrate that a new feature works as intended and that it is unit-tested.

**Comment by khotilov** [comment]
@mcuelenaere  I've looked into the details, and below are some observations:
- Your current code is based on the assumption that the `base_score` of an internal node corresponds to an average prediction value of all the leafs reachable from that node. However, this is not a correct assumption since, generally speaking, the `base_score` of an internal node is a prediction value only in the case when this node is pruned to be a leaf, and having children may easily change that value. In other words, instead of working with an actual tree as it is in a model, you are kind of mixing up contributions from a set of all possible pruned paths from that tree. To do it right, I don't see any easier way other than using the actual leaf predication values and the leaf and node cover values for calculating the average predictions of internal nodes.
- Picking `learning_rate` from a single tree is not good (`learning_rate` could be different in each tree). However, `learning_rate` would not be needed when operating with the actual leaf prediction values.
- For the multi-class classification setting, it's important to explain feature contributions separately for each class (and return a vector that is `num_output_group` times larger).

**Comment by mcuelenaere** [comment]
@khotilov 
  * Can you give me a counter-example where this assumption does not uphold? The code below calculates the leaf value as is done in the eli5 library and the results are pretty much exactly the same (the difference is possibly caused by a cumulation of rounding errors)
  * Done.
  * Done, but untested atm.


```python
import json
import numpy as np

def parent_value(children):
    covers = np.array([float(child['cover']) for child in children])
    covers /= np.sum(covers)
    leafs = np.array([child['leaf'] for child in children])
    return np.sum(leafs * covers)

def recursive_walker(parent):
    for child in parent['children']:
        if 'leaf' not in child:
            recursive_walker(child)

    if 'leaf' not in parent:
        parent['leaf'] = parent_value(parent['children'])

booster = ...

for raw_json in booster.get_dump(with_stats=True, dump_format='json'):
    learning_rate = 0.01
    try:
        a = json.loads(raw_json)
    except ValueError:
        print('INVALID JSON')
        continue
    recursive_walker(a)
    print(a['weight'] * learning_rate, a['leaf'])
```
```diff
diff --git a/src/tree/tree_model.cc b/src/tree/tree_model.cc
index c7878672..0e70d031 100644
--- a/src/tree/tree_model.cc
+++ b/src/tree/tree_model.cc
@@ -120,9 +120,10 @@ void DumpRegTree(std::stringstream& fo,  // NOLINT(*)
     if (with_stats) {
       if (format == "json") {
         fo << ", \"gain\": " << tree.stat(nid).loss_chg
-           << ", \"cover\": " << tree.stat(nid).sum_hess;
+           << ", \"cover\": " << tree.stat(nid).sum_hess
+           << ", \"weight\": " << tree.stat(nid).base_weight;
       } else {
-        fo << ",gain=" << tree.stat(nid).loss_chg << ",cover=" << tree.stat(nid).sum_hess;
+        fo << ",gain=" << tree.stat(nid).loss_chg << ",cover=" << tree.stat(nid).sum_hess << ",weight=" << tree.stat(nid).base_weight;
       }
     }
     if (format == "json") {
```

**Comment by khotilov** [comment]
@mcuelenaere Here's a simple example using a depth-two tree with learning rate 1:
```R
library(mlbench)
data(PimaIndiansDiabetes2)
dtrain <- xgb.DMatrix(as.matrix(PimaIndiansDiabetes2[, -9]),
                      label = as.numeric(PimaIndiansDiabetes2$diabetes)-1)
param <- list(max_depth = 2, eta = 1, nthread = 2, subsample = 1, 
              objective = "binary:logistic", eval_metric = "auc",
              base_score = mean(getinfo(dtrain,"label")))
set.seed(1)
bst <- xgb.train(param, dtrain, 1, verbose=2)
xgb.dump(bst, 'zz', with_stats = T, dump_format = "json")
```
which results in 
```json
[
  { "nodeid": 0, "depth": 0, "split": 1, "split_condition": 127.5, "yes": 1, "no": 2, "missing": 1,
    "gain": 137.63435363769531, "cover": 174.47915649414062, "weight": -8.8992997859804746e-008,
    "children": [
    { "nodeid": 1, "depth": 1, "split": 7, "split_condition": 28.5, "yes": 3, "no": 4, "missing": 3,
      "gain": 31.213098526000977, "cover": 110.18540954589844, "weight": -0.67675065994262695,
      "children": [
      { "nodeid": 3, "leaf": -1.1438477039337158, "cover": 61.567516326904297 },
      { "nodeid": 4, "leaf": -0.074108093976974487, "cover": 48.617893218994141 }
    ]},
    { "nodeid": 2, "depth": 1, "split": 5, "split_condition": 29.950000762939453, "yes": 5, "no": 6, "missing": 5,
      "gain": 39.552539825439453, "cover": 64.293754577636719, "weight": 1.1524039506912231,
      "children": [
      { "nodeid": 5, "leaf": -0.13800567388534546, "cover": 17.266166687011719 },
      { "nodeid": 6, "leaf": 1.6191864013671875, "cover": 47.027587890625 }
    ]}
  ]}
]
```
Note that I've increased the dump precision (`fo << std::setprecision(17);`).

By doing some simple manual calculations we get:
```R
> # nodes 3 & 4
> (61.567516326904297*(-1.1438477039337158) + 48.617893218994141 *(-0.074108093976974487))/(61.567516326904297 + 48.617893218994141)
[1] -0.6718389
> # nodes 5 & 6
> (17.266166687011719*(-0.13800567388534546) + 47.027587890625*1.6191864013671875)/(17.266166687011719 + 47.027587890625)
[1] 1.14729
```
These are non-negligibly different from the corresponding weights of nodes 1 and 2: -0.67675 and 1.15240. 

Note that this was just a random example of a depth-2 tree with fairly simple parameters settings and an uncomplicated dataset -  I didn't try to seek out an example with larger differences, but I saw some larger ones previously. Neither I've spent any time to figure out what drives those differences. And I cannot really tell why you have observed the same numbers. In your python code, however, try printing the numbers for all internal nodes, not just only the tree roots.

I would't rely on feature contributions based on the predictions from smaller intermediate trees (base weights of internal nodes), since I see that they do not necessarily correspond to averaged predictions derived from the leafs of grown trees in a final model, and the nature and magnitudes of the differences are not well understood.

**Comment by mcuelenaere** [comment]
@khotilov I reworked the feature contribution calculation code to (lazily) calculate the node mean values and use those instead of the base_weight stats. I also added 1 simple test which just compares the sum of the contributions with the actual prediction (which actually revealed a couple of bugs, that I fixed).

**Comment by goessling** [comment]
@khotilov and @mcuelenaere: The discrepancies you're seeing are due to the L2 regularization. If you account for that then you get identical results for the base weight of internal nodes and the averaged predictions derived from the leaves (up to rounding errors):
```python
> reg_lambda = 1.0  # default if not set explicitly
> # nodes 3 & 4
> ((61.567516326904297+reg_lambda)*(-1.1438477039337158) + (48.617893218994141+reg_lambda) *(-0.074108093976974487))/(61.567516326904297 + 48.617893218994141 + reg_lambda)
[1] -0.676750642843256
> # nodes 5 & 6
> ((17.266166687011719+reg_lambda)*(-0.13800567388534546) + (47.027587890625+reg_lambda)*1.6191864013671875)/(17.266166687011719 + 47.027587890625 + reg_lambda)
[1] 1.1524039787075722
```
Without accounting for the L2 regularization the differences will be large if there are few samples per node (small sum_hess) and they will be small if there are many samples per node.
I would argue that the base weights are the more appropriate quantity to use because they incorporate the L2 regularization (it is the value that would have been used if that node had become a leaf) and they are more easily accessible.

**Comment by khotilov** [comment]
@goessling Nice catch of the role of regularization in this issue! Note that L1 regularization could be used as well, and it would also shrink children's weights... 

However, I disagree about the appropriateness of base weights. The approach we are discussing here is based on probabilistic interpretation of data flowing down the trees and how it relates to an individual case prediction. Predictions are done by a final fully grown model, not by some intermediate "would have been" trees. Regularization parameters help to build the final tree structures. But once the trees are built, prediction does not care about what were the regularizations or about any other details of the construction process - splits and leaf values are all what really matters.  The stored hessian sums provide a reasonable picture of the probabilistic data-flow (which, BTW, could also be fine-tuned by using the "update" process #1670 ) that is simple to make use of.

File: include/xgboost/tree_model.h [hunk 1]
```diff
@@ -535,6 +561,56 @@ inline bst_float RegTree::Predict(const RegTree::FVec& feat, unsigned root_id) c
   return (*this)[pid].leaf_value();
 }
 
+inline void RegTree::FillNodeMeanValues() {
+  if (this->node_mean_values.size() > 0) {
```
Comment (khotilov) [inline include/xgboost/tree_model.h]: this needs to be compared to param.num_nodes

File: include/xgboost/tree_model.h [hunk 2]
```diff
@@ -535,6 +561,56 @@ inline bst_float RegTree::Predict(const RegTree::FVec& feat, unsigned root_id) c
...
+    return;
+  }
+  this->node_mean_values.resize(this->param.num_nodes);
+  int root_id = 0;
+  this->FillNodeMeanValue(root_id);
```
Comment (khotilov) [inline include/xgboost/tree_model.h]: this needs to be done for all param.num_roots root nodes

**Comment by goessling** [comment]
@khotilov I can definitely see the motivation for using averages of leaf values, which avoids using regularized values at the inner nodes. I think one argument in favor of using the `base_weights` is that the computation for the regular feature importance score `get_score(importance_type='gain')` is also based on regularized values (both L1 and L2), see https://github.com/dmlc/xgboost/blob/master/src/tree/param.h#L227. As a result, with `base_weights` the contribution of one feature (pooled over all training examples that pass through that node) would correspond exactly to the gain of that feature according to the importance score. With the averages of leaf values this correspondence would no longer hold. Or in other words, if the effect of a feature should be based on a difference of leaf value averages (instead of a difference of regularized averages) then the same applies to the feature importance score and it should be updated accordingly I think.

**Comment by khotilov** [comment]
@goessling To me that's simply an argument for having one more type of feature importance.

**Comment by mcuelenaere** [comment]
@khotilov @goessling So how do we proceed here? Is the current approach the way to go or do we want to support both average leaf values and regularized values?

**Comment by khotilov** [comment]
@mcuelenaere The averaged values approach is more valid. Think about this way: when explaining predictions of a model that has a structure of linear regression, we don't want to explain the process of how that model was trained or how its coefficients were evolving during the training (e.g., whether there were any kinds of regularizations involved or not, or whether it was Newton-Raphson, SGD, boosting, etc.) - all we need are just the final values of linear terms. The same thing holds for tree structure-based models: once the trees are build and the prediction procedure is defined, we can explain them in a probabilistic manner, without any need of knowing of how exactly they were built.

I've done some cross-checks with my script yesterday, and the results were in pretty good agreement! One minor thing that constantly was throwing me off, however, was that the bias was not the first but the last column. Maybe it's just me, but I find such representation as rather unusual.

Also, here you only have an option for dense output, which would be rather awkward for working with very wide sparse data. So I was trying to see whether there might be any different design decisions needed to accommodate possible further extensions, e.g., to output sparse contributions or something like "top-N" most contributing features. So far I didn't really think this through. Do you have any thoughts on this?

File: src/gbm/gbtree.cc [hunk 3]
```diff
@@ -537,6 +545,62 @@ class GBTree : public GradientBooster {
...
+            }
+            trees[j]->CalculateContributions(feats, root_id, p_contribs);
+          }
+          feats.Drop(batch[i]);
+          // add base margin to BIAS feature
+          if (base_margin.size() != 0) {
+            p_contribs[ncolumns - 1] += base_margin[row_idx];
+          } else {
+            p_contribs[ncolumns - 1] += base_margin_;
+          }
```
Comment (khotilov) [inline src/gbm/gbtree.cc]: p_contribs[ncolumns - 1] +=  (base_margin.size() != 0) ?
  base_margin[row_idx * mparam.num_output_group + gid] : base_margin_;

**Comment by mcuelenaere** [comment]
@khotilov I don't have a lot of time I can dedicate to this task and I'd like to see this upstreamed.

> One minor thing that constantly was throwing me off, however, was that the bias was not the first but the last column. Maybe it's just me, but I find such representation as rather unusual.

Sure, I don't mind it being first or last; as long as it is consistent and documented.

> Also, here you only have an option for dense output, which would be rather awkward for working with very wide sparse data. So I was trying to see whether there might be any different design decisions needed to accommodate possible further extensions, e.g., to output sparse contributions or something like "top-N" most contributing features. So far I didn't really think this through. Do you have any thoughts on this?

Can we for now throw an error when passing sparse data to the function and keep track of this in a separate GitHub issue? I'd rather not expand the scope of this too much, as that would make upstreaming this take a lot longer.

Instead of storing the feature contributions in a fixed-size collection, we could look into using a dynamically-sized one (with an upper bound). How to pass this to the caller/convert this into a sparse dataset, I have not looked into though.

**Comment by khotilov** [comment]
@mcuelenaere Since it is usually the case (even in sparse settings) that only a rather manageable set of most important features would be contributing any appreciable amounts to predictions, what do you think about an interface that would allow a user to pass a vector of select feature indices as a parameter? The returned contributions would be for these requested features, a combined contribution of any remaining features, and, separately, the bias; and all the contributions would be returned in case when that vector was left empty. That could be a rather practical interface option. It is fairly simple to implement as well. The only thing is that it might need a separate API method, as `Predict` is becoming increasingly crowded...

@tqchen what is your opinion on this?

**Comment by tqchen** [comment]
Yes, I think as long as the API is clean, we can consider merge this in

**Comment by khotilov** [comment]
Thanks for the feedback @tqchen 

@mcuelenaere could you please fix the `base_margin` handling in multiclass setting that I've pointed earlier? Then we can merge this PR.

**Comment by mcuelenaere** [comment]
@khotilov I rebased on master + changed the `base_margin` handling. Sorry for the delay, I haven't had much time to spend on this. What's the plan exactly now? Does this need to be reworked into a different API or will it go in as-is?

**Comment by khotilov** [comment]
@mcuelenaere Thanks! 

The current changes to API are fairly minor, so we can merge it in as an "experimental feature". Will keep looking into what would be the best approach for high number of sparse features though.

I will shortly add the R binding, some more tests, and the gblinear implementation.
